### PR TITLE
Add support for Bech32 Bitcoin addresses

### LINKED ISSUE #734: No support for Bech32 (bc1) addresses
## Problem Definition

I have some BTC balance in a Bech32 address (those that begin with `bc1`).
blockchain.info doesn't support those on the API level.
For example, here is a random address I found on blockchain.info: `bc1qhkje0xfvhmgk6mvanxwy09n45df03tj3h3jtnf`
This can be viewed in the interface: <https://www.blockchain.com/btc/address/bc1qhkje0xfvhmgk6mvanxwy09n45df03tj3h3jtnf>
But the API does not support it: <https://blockchain.info/q/addressbalance/bc1qhkje0xfvhmgk6mvanxwy09n45df03tj3h3jtnf>

The relevant rotkehlchen.log entry is:

```
20/02/2020 14:57:07 CET -- DEBUG:rotkehlchen.api.rest:Request successful response={"result": {"status": "completed", "outcome": {"result": null, "message": "The given string bc1qsvhnjfuhfund4684tylcqnd402jvdr9lt3h7l0 is not a valid BTC address"}}, "message": ""}, status_code=200
```

Maybe the `query_btc_account_balance` function in `blockchain.py`  could be modified to use different APIs depending on address type:

```py
        try:
            if account[0:3] == 'bc1':
                btc_resp = request_get(
                    url='https://api.blockcypher.com/v1/btc/main/addrs/%s/balance' % account,
                )['balance']
            else:
                btc_resp = request_get_direct(
                    url='https://blockchain.info/q/addressbalance/%s' % account,
                    handle_429=True,
                    # If we get a 429 then their docs suggest 10 seconds
                    # https://blockchain.info/q
                    backoff_in_seconds=10,
                )
```

### System Description

```
uname -a
command -v geth && geth version
command -v parity && parity --version
command -v rotkehlchen && rotkehlchen version
[ -d .git ] && git rev-parse HEAD
```
gives
```
Linux alladin 5.4.6-arch3-1 #1 SMP PREEMPT Tue, 24 Dec 2019 04:36:53 +0000 x86_64 GNU/Linux
/usr/bin/geth
Geth
Version: 1.9.9-stable
Git Commit: 017449971e1e9e220efcd97d3313a0e27f47003b
Git Commit Date: 20191206
Architecture: amd64
Protocol Versions: [64 63]
Go Version: go1.13.5
Operating System: linux
GOPATH=
GOROOT=/usr/lib/go
/home/hjort/.local/bin/rotkehlchen
1.1.2.dev0+g42481b5.d20200220
42481b5f2ef28f25c99663a092f5c85157285379
```

### PR DESCRIPTION
**Pull request by hjorthjort**
Fixes #734.
Adds a check for whether an address is a Bech32 address, and declares such addresses valid in addition to base58 addresses.
Uses a new endpoint (blockcypher.com) to query balances for Bech32 addresses, since blockchain.info does not support querying them through the API.

# Manual testing checklist

New User

   - [X] Create a new user and see that it works. Both with and without a premium key. With a premium key make sure that you can verify that pulling data from the server works.
   - [X] Provide mismatching passwords and see it’s handled properly.
   - [x] Provide wrong premium keys and see it’s handled properly

Sign in existing user

   - [X] Sign in an existing user with a wrong password and see it’s handled.
   - [X] Sign in a non-existing user and see it’s handled
   - [X] Sing in an existing user and see it works // TYPO

External Trades

   - [x] Add an external trade and see it’s added in the table
   - [x] Edit an external trade from the table and see it’s altered
   - [x] Delete an external trade from the table and see it’s removed
   - [x] Expand the details on a trade and see they are shown properly

Data Importing

   - [ ] Import some data from cointracking.info and see that works properly

Exchanges

   - [x] Add an invalid exchange API key and see it’s handled properly
   - [x] Add a valid exchange API key and see it works. See that dashboard balances are also updated.
   - [x] Remove an exchange and see that it works and that the dasboard balances are updated.

External Services

   - [x] Add an API key for all external services
   - [x] Remove an API key for all external services

Application and Accounting Settings

   - [x] Change all application settings one by one and see the changes are reflected.
   - [X] Same as above but for invalid values (if possible) and see they are handled.
   - [X] Change the profit currency and see it works
   - [X] Change all accounting settings one by one and see the changes are reflected.
   - [X] Same as above but for invalid values (if possible) and see they are handled.

Accounts and Balances
Fiat

   - [X] Add a fiat balance and see it works
   - [X] Remove a fiat balance and see it works // Remove => Modify
   - [X] See that adding non number or negative is handled

Ethereum Accounts

   - [X] Add an ethereum account and see it works
   - [X] Add an invalid ethereum account and see it is handled properly
   - [X] Remove an ethereum account and see it works
   - [X] After adding tokens to an account that has it expand the account and see all tokens owned by it are shown.

Ethereum Tokens

   - [x] Track an ethereum token and see it works. Works is defined as being added:
   - [x]         In the dashboard
   - [x]         In the owned tokens
   - [x]         In total blockchain balances
   - [x]         In the expanded asset details of ETH accounts that own it.

   - [x] Remove an ethereum token and see it works. Works means being removed from all the above.

Bitcoin accounts

   - [X] Add a bitcoin account and see it works
   - [X] Add an invalid bitcoin account and see it is handled properly
   - [X] Remove a bitcoin account and see it works

Tax Report

   - [x] Check that invalid input in the date range are handled properly
   - [x] Create a big tax report over many exchanges for a long period of time and see that it’s correct and no unexpected problems occur.
   - [x] Create a CSV export of the report and see it works

Premium Analytics

   - [ ] Check they work for a premium account
   - [ ] Modify the range of the netvalue graph and see it works properly
   - [ ] Change the asset and modify the range of the graph of amount and value of an asset and see it works properly
   - [ ] Check the netvalue distribution by location works properly
   - [ ] Check the netvalue distribution by asset works properly and that you can modify the number of assets shown in the graph

### CODE REVIEW

**Comment by hjorthjort** [comment]
There, all manual tests I have access to are done. I don't have a cointracking account or a premium key (yet), so I'll leave those tests for now.

Seems Travis is failing due to an exchange having updated its listings, so I'm assuming it's unrelated.

File: rotkehlchen/chain/manager.py [hunk 1]
```diff
@@ -203,13 +203,18 @@ def query_btc_account_balance(account: BTCAddress) -> FVal:
...
         """
         try:
-            btc_resp = request_get_direct(
-                url='https://blockchain.info/q/addressbalance/%s' % account,
-                handle_429=True,
-                # If we get a 429 then their docs suggest 10 seconds
-                # https://blockchain.info/q
-                backoff_in_seconds=10,
-            )
+            if account[0:3] == 'bc1':
```
Comment (hjorthjort) [inline rotkehlchen/chain/manager.py]: This could have used the new `bech32` requirement, but I personally prefer to minimize dependency surface when it's easy to do.

**Comment by LefterisJP** [comment]
@hjorthjort 

>Seems Travis is failing due to an exchange having updated its listings, so I'm assuming it's unrelated.

It's completely unrelated. Linux passing and OSX failing with socket or other timeouts is due to Travis OSX VMs being completely unreliable. It's one of the reasons we want to switch to something else eventually.

**Comment by LefterisJP** [comment]
Hey @hjorthjort it's super awesome you tried to do the manual test checklist (which we only have for the frontend, since it's not well covered in tests -- the backend is).

But this is mostly for creating releases. Definitely not a requirement for new contributors to check them but it's awesome you did

**Comment by hjorthjort** [comment]
Ah, cool. I figured that was probably the case, but it was a good way to
try out the app functionality anyway.

Lefteris Karapetsas <notifications@github.com> schrieb am Mo., 16. März
2020, 3:26 PM:

File: rotkehlchen/chain/bitcoin.py [hunk 2]
```diff
@@ -1,10 +1,24 @@
...
+import bech32
 
 
 def is_valid_btc_address(value: str) -> bool:
-    """Validates a bitcoin address for the mainnet
+    return is_valid_base58_address(value) or is_valid_bech32_address(value)
+
+
+def is_valid_bech32_address(value: str) -> bool:
+    """Validates a bitcoin base58 address for the mainnet
```
Comment (LefterisJP) [inline rotkehlchen/chain/bitcoin.py]: This doscstring should say bech32

File: rotkehlchen/chain/manager.py (hunk 1 above)
Comment (LefterisJP) [reply rotkehlchen/chain/manager.py]: Looking at [BIP0173](https://en.bitcoin.it/wiki/BIP_0173) it's always going to be starting with `bc1`, right? So this is a safe check covering all edge cases?

File: rotkehlchen/tests/test_blockchain.py [hunk 3]
```diff
@@ -109,6 +109,12 @@ def test_is_valid_btc_address():
...
     assert is_valid_btc_address(UNIT_BTC_ADDRESS3)
     assert is_valid_btc_address('18ddjB7HWTVxzvTbLp1nWvaBxU3U2oTZF2')
+    assert is_valid_btc_address('bc1qhkje0xfvhmgk6mvanxwy09n45df03tj3h3jtnf')
+    assert is_valid_btc_address('bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4')
+    assert is_valid_btc_address('bc1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3qccfmv3')
 
     assert not is_valid_btc_address('foo')
     assert not is_valid_btc_address('18ddjB7HWTaxzvTbLp1nWvaixU3U2oTZ1')
+    assert not is_valid_btc_address(
+        'tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3qccfmv3')
```
Comment (LefterisJP) [inline rotkehlchen/tests/test_blockchain.py]: This is bech32 testnet, right?

File: rotkehlchen/chain/bitcoin.py (hunk 2 above)
Comment (hjorthjort) [reply rotkehlchen/chain/bitcoin.py]:
```suggestion
    """Validates a bitcoin Bech32 address for the mainnet
```

File: rotkehlchen/chain/manager.py (hunk 1 above)
Comment (hjorthjort) [reply rotkehlchen/chain/manager.py]: Yes, that's how I understand it. Different chains use different human-readable parts, and `bc1` is the one only for Bitcoin mainnet.

Looking at that link, there are lots and lots of positive and negative test cases. Should we add some more?

File: rotkehlchen/tests/test_blockchain.py (hunk 3 above)
Comment (hjorthjort) [reply rotkehlchen/tests/test_blockchain.py]: From the BIP:

>  MUST verify that the human-readable part is "bc" for mainnet and "tb" for testnet.

File: rotkehlchen/chain/manager.py (hunk 1 above)
Comment (LefterisJP) [reply rotkehlchen/chain/manager.py]: Sure, more tests are always welcome

**Comment by hjorthjort** [comment]
Good thing we added more test cases, turns out we can be even more stringent about what addresses we accept.

**Comment by LefterisJP** [comment]
@hjorthjort nice! Can you also add the changelog entry as requested above?

File: rotkehlchen/chain/manager.py (hunk 1 above)
Comment (hjorthjort) [reply rotkehlchen/chain/manager.py]: A finding: The tests show that addresses MAY be uppercase, but lowercase is the standard format. The API only accepts lowercase addresses, so we need to lowercase them when querying.

Comment (LefterisJP) [reply rotkehlchen/chain/manager.py]: When querying what?

**Comment by LefterisJP** [comment]
@hjorthjort Can you rebase on develop or should I do that for you? Also is there anything left to do here, regarding tests? I would like to have this merged ASAP.

**Comment by hjorthjort** [comment]
Go ahead and rebase! And I think that's it.

Lefteris Karapetsas <notifications@github.com> schrieb am Do., 19. März
2020, 12:34 PM:

**Comment by LefterisJP** [comment]
@hjorthjort There seem to be some linting failures too but I will handle them at rebase.
